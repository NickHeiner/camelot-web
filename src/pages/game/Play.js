import React, {PureComponent} from 'react';
import autobind from 'autobind-decorator';
import Avatar from '../../components/Avatar';
import Board from '../../components/Board';
import './Play.less';
import {Button} from 'react-bootstrap';
import CapturedPieces from '../../components/CapturedPieces';
import _ from 'lodash';
import {firebaseConnect} from 'react-redux-firebase';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {makeMove} from '../../actions';

import camelotEngine from 'camelot-engine';
const {isValidMove, getGameWinner} = camelotEngine().query();

export class PresentationGamePlay extends PureComponent {
  render() {
    let gameDisplay;
    if (this.props.host === undefined || this.props.opponent === undefined) {
      gameDisplay = <p>Loading...</p>;
    } else if (this.props.game === null) {
      gameDisplay = <p>This link is not valid. Did someone share it with you incorrectly?</p>;
    } else {
      const currentUserUid = this.props.currentUser.uid,
        currentUserIsHost = currentUserUid === this.props.host.uid;

      let findOpponentMessage;
      if (!this.props.opponent) {
        if (currentUserIsHost) {
          findOpponentMessage = <p>Find someone to play with you by sharing this link with them.</p>;
        } else {
          findOpponentMessage = (
              <div>
                <p>This game is currently looking for an opponent.</p>
                <Button bsStyle="primary" onClick={this.joinGame}>Join</Button>
              </div>
          );
        }
      }

      const gameState = this.props.game.gameState;
      let activeUser = null;
      let isCurrentUserActive = false;
      let currentUserPlayer = null;
      let userHasValidMove = false;
      let gameWinner = null;
      if (gameState) {
        activeUser = gameState.turnCount % 2 === 0 ? 'host' : 'opponent';

        isCurrentUserActive = currentUserUid === this.props[activeUser].uid;

        const opponentUid = this.props.opponent && this.props.opponent.uid;

        currentUserPlayer = currentUserUid === this.props.host.uid ? 'playerA' 
                    : currentUserUid === opponentUid ? 'playerB' : null;

        // I don't know why isValidMove considers [] and [singleMove] to be valid.
        userHasValidMove = this.props.chosenMoveSteps.size > 1 && 
                    isValidMove(gameState, this.props.chosenMoveSteps, currentUserPlayer);

        gameWinner = getGameWinner(gameState);
      }

      gameDisplay = (
                <div>
                    <Board gameState={gameState} 
                        isCurrentUserActive={isCurrentUserActive}
                        possibleMove={this.props.chosenMoveSteps}
                        gameId={this.props.gameId}
                        message={findOpponentMessage || this.getWinMessage(gameWinner)}
                        currentUserPlayer={currentUserPlayer} />
                    <div className="control-bar">
                        <Avatar currentUser={this.props.host} isActive={activeUser === 'host'} />
                        <CapturedPieces whosePiecesWereCaptured="opponent" gameState={gameState} />
                        {
                            isCurrentUserActive
                                ? <Button bsStyle="primary" 
                                    disabled={!userHasValidMove} 
                                    onClick={this.makeMove}>Make Move</Button>
                                : <p>Other player's turn</p>
                        }
                        <CapturedPieces whosePiecesWereCaptured="host" gameState={gameState} />
                        <Avatar currentUser={this.props.opponent} isActive={activeUser === 'opponent'} />
                    </div>
                </div>
            );
    }

    return gameDisplay;
  }

  @autobind
  joinGame() {
    this.props.firebase
            .ref(`/games/${this.props.gameId}/opponent`)
            .set(this.props.currentUser.uid);
  }

  @autobind
  makeMove() {
    this.props.makeMove(this.props.gameId, this.props.game.gameState, this.props.chosenMoveSteps);
  }

  @autobind
  getWinMessage(gameWinner) {
    if (!gameWinner) {
      return;
    }
    const winningUser = gameWinner === 'playerA' ? this.props.host : this.props.opponent;
    return `${winningUser.displayName} wins!`;
  }
}

@firebaseConnect(['/games', '/users'])
@connect(
  ({firebase, ui}, ownProps) => {
    const game = _.get(firebase.data, ['games', ownProps.params.id]);
    
    return {
      game,
      chosenMoveSteps: ui.get('chosenMoveSteps'),
      host: game && _.get(firebase.data.users, game.host),
      opponent: game && _.get(firebase.data.users, game.opponent, null),
      currentUser: firebase.profile
    };
  },
  dispatch => bindActionCreators({
    makeMove
  }, dispatch)
)
export default class GamePlayContainer extends PureComponent {
  render = () => Boolean(this.props.game) && 
    <PresentationGamePlay gameId={this.props.params.id} {...this.props} />
}
