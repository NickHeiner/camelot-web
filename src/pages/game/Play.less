.control-bar {
    @spacing: 10px;
    position: absolute;
    bottom: 0;
    width: 100%;
    display: flex;
    justify-content: space-between;
    padding: @spacing;

    background-color: #ccc;
}

.host {
    color: white;
}
.opponent {
    color: black;
}

.captured {
    display: flex;
    flex-direction: column;
}
