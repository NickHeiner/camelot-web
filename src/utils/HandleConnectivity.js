import {connect} from 'react-redux';
import React, {PureComponent} from 'react';
import _ from 'lodash';

export default (OnlineContainer, OfflineContainer) => {
    @connect(
        state => _.pick(state, 'offline')
    )
    class HandleConnectivity extends PureComponent {
        render = () => this.props.offline ? <OfflineContainer /> : <OnlineContainer />;
    };
    
    return HandleConnectivity;
};
