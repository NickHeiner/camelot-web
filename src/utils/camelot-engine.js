import {Iterable} from 'immutable';
// This is a bit hacky but w/e.
import pairwise from 'camelot-engine/lib/util/pairwise';
import camelotEngine from 'camelot-engine';
const query = camelotEngine().query();

const ensureJS = val => Iterable.isIterable(val) ? val.toJS() : val;

// TODO clean this up by always using rest params
const proxyWithJsArgs = fn => (...args) => fn(...args.map(ensureJS));

export const isValidMove = (gameState, possibleMoveSteps, movingPlayerName) =>
  query.isValidMove(ensureJS(gameState), ensureJS(possibleMoveSteps), movingPlayerName);

export const getBoardSpace = (...args) =>
  query.getBoardSpace(...args.map(ensureJS));

export const isGoal = (gameState, ...rest) =>
  query.isGoal(ensureJS(gameState), ...rest);

export const getPairs = proxyWithJsArgs(pairwise);
export const getCoordsBetween = proxyWithJsArgs(query.getCoordsBetween);
