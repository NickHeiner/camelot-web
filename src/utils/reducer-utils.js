
export const getCurrentGame = (fullState, gameId) => fullState.firebase.getIn(['data', 'games', gameId]);

export const getCurrentUserPlayerName = (fullState, gameId) => {
  const currentUserUid = fullState.firebase.getIn(['profile', 'uid']);
  const currentGame = getCurrentGame(fullState, gameId);
  const currentHostUid = fullState.firebase.getIn(['data', 'users', currentGame.get('host'), 'uid']);
  const currentOpponentUid = fullState.firebase.getIn(['data', 'users', currentGame.get('opponent'), 'uid']);

  if (currentUserUid === currentHostUid) {
    return 'playerA';
  }
  if (currentUserUid === currentOpponentUid) {
    return 'playerB';
  }
  return null;
};
