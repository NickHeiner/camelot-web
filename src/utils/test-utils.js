import ReactShallowRenderer from 'react-test-renderer/shallow';
import camelotEngine from 'camelot-engine';
const {getBoardSpace} = camelotEngine().query();

export const getDummyCurrentUser = () => ({
  photoURL: 'photoURL',
  displayName: 'Display Name'
});

// TODO: add method to set a piece's location directly
// without breaking encapsulation like we're doing here.
export const setPieceInBoardSpace = (gameState, {row, col}, piece) => {
  getBoardSpace(gameState, {row, col}).piece = {
    type: 'pawn',
    player: 'playerA'
  };
  return gameState;
};

export const getGameStateWithPiece = (coords, piece) => setPieceInBoardSpace(
  camelotEngine().createEmptyGame(),
  coords, piece
);

export const shallowRenderComponent = component => {
  const renderer = new ReactShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
};
