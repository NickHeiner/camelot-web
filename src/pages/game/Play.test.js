import {PresentationGamePlay} from './Play';
import React from 'React';
import {fromJS} from 'immutable';
import {setPieceInBoardSpace, getGameStateWithPiece, shallowRenderComponent} from '../../utils/test-utils';

describe('GamePlay', () => {
  it('renders with no games', () => {
    expect(shallowRenderComponent(
      <PresentationGamePlay />
    )).toMatchSnapshot();
  });

  it('renders with a game that has a winner', () => {
    const gameState = getGameStateWithPiece({row: 0, col: 5}, {
      type: 'pawn',
      player: 'playerA'
    });
    const gameStateWithWinner = setPieceInBoardSpace(gameState, {row: 0, col: 6}, {
      type: 'pawn',
      player: 'playerA'
    });

    const game = fromJS({
      gameState: gameStateWithWinner
    });

    const host = fromJS({
      uid: 'host-uid'
    });

    const opponent = fromJS({
      uid: 'host-uid'
    });

    const chosenMoveSteps = fromJS([]);

    expect(shallowRenderComponent(
      <PresentationGamePlay game={game} host={host} 
        opponent={opponent} currentUser={host} chosenMoveSteps={chosenMoveSteps} />
    )).toMatchSnapshot();
  });
});
