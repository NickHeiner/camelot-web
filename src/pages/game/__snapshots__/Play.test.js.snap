// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePlay renders with a game that has a winner 1`] = `
<div>
  <Board
    currentUserPlayer={null}
    gameId={undefined}
    gameState={undefined}
    isCurrentUserActive={false}
    message={undefined}
    possibleMove={Immutable.List []}
  />
  <div
    className="control-bar"
  >
    <Avatar
      currentUser={
        Immutable.Map {
          "uid": "host-uid",
        }
      }
      isActive={false}
    />
    <CapturedPieces
      gameState={undefined}
      whosePiecesWereCaptured="opponent"
    />
    <p>
      Other player's turn
    </p>
    <CapturedPieces
      gameState={undefined}
      whosePiecesWereCaptured="host"
    />
    <Avatar
      currentUser={
        Immutable.Map {
          "uid": "host-uid",
        }
      }
      isActive={false}
    />
  </div>
</div>
`;

exports[`GamePlay renders with no games 1`] = `
<p>
  Loading...
</p>
`;
